File: packages/python-sdk/strait/composition/_cost_budget.py
"""Cost budget tracking and enforcement."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, TypeVar

from strait._errors import CostBudgetExceededError

T = TypeVar("T")


@dataclass
class CostBudgetOptions:
    max_cost_microusd: int
    on_warning: Callable[[int, int], None] | None = None
    warning_threshold: float = 0.8


class CostTracker:
    def __init__(self, options: CostBudgetOptions) -> None:
        self._current_cost = 0
        self._options = options
        self._warning_fired = False

    def add(self, cost_microusd: int) -> None:
        self._current_cost += cost_microusd

        if (
            not self._warning_fired
            and self._options.on_warning is not None
            and self._current_cost >= self._options.max_cost_microusd * self._options.warning_threshold
        ):
            self._warning_fired = True
            self._options.on_warning(self._current_cost, self._options.max_cost_microusd)

        if self._current_cost >= self._options.max_cost_microusd:
            raise CostBudgetExceededError(
                f"Cost budget exceeded: {self._current_cost} >= "
                f"{self._options.max_cost_microusd} microusd",
                current_cost_microusd=self._current_cost,
                max_cost_microusd=self._options.max_cost_microusd,
            )

    def current(self) -> int:
        return self._current_cost

    def remaining(self) -> int:
        return max(0, self._options.max_cost_microusd - self._current_cost)

    def is_exceeded(self) -> bool:
        return self._current_cost >= self._options.max_cost_microusd


def create_cost_tracker(options: CostBudgetOptions) -> CostTracker:
    return CostTracker(options)


async def with_cost_budget(
    fn: Callable[[CostTracker], Awaitable[T]],
    options: CostBudgetOptions,
) -> T:
    tracker = create_cost_tracker(options)
    return await fn(tracker)

File: packages/ruby-sdk/lib/strait/composition/cost_budget.rb
# frozen_string_literal: true

module Strait
  module Composition
    # Options for cost budget tracking.
    CostBudgetOptions = Struct.new(:max_cost_microusd, :on_warning, :warning_threshold, keyword_init: true) do
      def initialize(**)
        super
        self.warning_threshold ||= 0.8
      end
    end

    # Tracks accumulated costs against a budget.
    class CostTracker
      def initialize(options)
        @current_cost = 0
        @options = options
        @warning_fired = false
      end

      def add(cost_microusd)
        @current_cost += cost_microusd

        threshold = (@options.max_cost_microusd * @options.warning_threshold).to_i
        if !@warning_fired && @options.on_warning && @current_cost >= threshold
          @warning_fired = true
          @options.on_warning.call(@current_cost, @options.max_cost_microusd)
        end

        if @current_cost >= @options.max_cost_microusd
          raise Strait::CostBudgetExceededError.new(
            "Cost budget exceeded: #{@current_cost} >= #{@options.max_cost_microusd} microusd",
            current_cost_microusd: @current_cost,
            max_cost_microusd: @options.max_cost_microusd
          )
        end
      end

      def current
        @current_cost
      end

      def remaining
        [@options.max_cost_microusd - @current_cost, 0].max
      end

      def exceeded?
        @current_cost >= @options.max_cost_microusd
      end
    end

    def self.create_cost_tracker(options)
      CostTracker.new(options)
    end

    def self.with_cost_budget(options, &block)
      tracker = create_cost_tracker(options)
      block.call(tracker)
    end
  end
end
